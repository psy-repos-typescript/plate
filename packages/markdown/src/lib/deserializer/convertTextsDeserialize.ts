import { getPluginType } from 'platejs';

import type { MdDelete, MdEmphasis, MdStrong } from '../mdast';
import type { MdDecoration } from '../types';
import type { DeserializeMdOptions } from './deserializeMd';

import { mdastToPlate } from '../types';
import { buildSlateNode } from './convertNodesDeserialize';

export const convertTextsDeserialize = (
  mdastNode: MdDelete | MdEmphasis | MdStrong,
  deco: MdDecoration,
  options: DeserializeMdOptions
) => {
  return mdastNode.children.reduce((acc: any, n: any) => {
    const key = mdastToPlate(options.editor!, mdastNode.type)
    const type = getPluginType(options.editor!, key)

    acc.push(
      ...buildSlateNode(
        n,
        { ...deco, [type]: true },
        options
      )
    );
    return acc;
  }, []);
};
