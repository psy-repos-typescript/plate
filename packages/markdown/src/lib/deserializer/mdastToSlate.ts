import type { Root } from 'mdast';

import { type Descendant, getPluginKey, KEYS } from 'platejs';

import type { MdRoot } from '../mdast';
import type { DeserializeMdOptions } from './deserializeMd';

import { convertNodesDeserialize } from './convertNodesDeserialize';

export const mdastToSlate = (
  node: Root,
  options: DeserializeMdOptions
): Descendant[] => {
  return buildSlateRoot(node, options);
};

const buildSlateRoot = (
  root: MdRoot,
  options: DeserializeMdOptions
): Descendant[] => {
  if (!options.splitLineBreaks) {
    root.children = root.children.map((child) => {
      if (child.type === 'html' && child.value === '<br />') {
        return {
          children: [{ type: 'text', value: '\n' }],
          type: 'paragraph',
        };
      }
      return child;
    });
    return convertNodesDeserialize(root.children, {}, options);
  }

  // Split line breaks into separate paragraphs
  const results: Descendant[] = [];
  let startLine = root.position?.start.line ?? 1;

  const addEmptyParagraphs = (count: number) => {
    if (count > 0) {
      results.push(
        ...Array.from({ length: count }).map(() => ({
          children: [{ text: '' }],
          type: options.editor
            ? (getPluginKey(options.editor, KEYS.p) ?? KEYS.p)
            : KEYS.p,
        }))
      );
    }
  };

  root.children?.forEach((child, index) => {
    const isFirstChild = index === 0;
    const isLastChild = index === root.children!.length - 1;

    if (child.position) {
      const emptyLinesBefore =
        child.position.start.line - (isFirstChild ? startLine : startLine + 1);
      addEmptyParagraphs(emptyLinesBefore);

      const transformValue = convertNodesDeserialize([child], {}, options);
      results.push(...transformValue);

      if (isLastChild) {
        const emptyLinesAfter =
          root.position!.end.line - child.position.end.line - 1;
        addEmptyParagraphs(emptyLinesAfter);
      }

      startLine = child.position.end.line;
    } else {
      const transformValue = convertNodesDeserialize([child], {}, options);
      results.push(...transformValue);
    }
  });

  return results;
};
