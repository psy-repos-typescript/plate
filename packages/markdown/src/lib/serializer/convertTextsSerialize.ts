import { type TText, getPluginType } from 'platejs';

import type { MdMark } from '../types';
import type { SerializeMdOptions } from './serializeMd';

import { getCustomMark } from './utils';
import { getSerializerByKey } from './utils/getSerializerByKey';

// inlineCode should be last because of the spec in mdast
// https://github.com/inokawa/remark-slate-transformer/issues/145
export const basicMarkdownMarks = ['italic', 'bold', 'strikethrough', 'code'];

export const convertTextsSerialize = (
  slateTexts: readonly TText[],
  options: SerializeMdOptions,
  key?: string
): MdMark[] => {
  const customLeaf: string[] = getCustomMark(options);

  const mdastTexts: MdMark[] = [];

  const starts: string[] = [];
  let ends: string[] = [];

  let textTemp = '';
  for (let j = 0; j < slateTexts.length; j++) {
    const cur = slateTexts[j]!;
    textTemp += cur.text;

    const prevStarts = starts.slice();
    const prevEnds = ends.slice();

    const prev = slateTexts[j - 1];
    const next = slateTexts[j + 1];
    ends = [];
    (
      [
        ...basicMarkdownMarks,
        // exclude repeated marks
        ...customLeaf.filter((k) => !basicMarkdownMarks.includes(k)),
      ] as const
    ).forEach((key) => {
      const nodeType = getPluginType(options.editor!, key);

      if (cur[nodeType]) {
        if (!prev?.[nodeType]) {
          starts.push(key);
        }
        if (!next?.[nodeType]) {
          ends.push(key);
        }
      }
    });

    const endsToRemove = starts.reduce<{ key: string; index: number }[]>(
      (acc, k, kIndex) => {
        if (ends.includes(k)) {
          acc.push({ key: k, index: kIndex });
        }
        return acc;
      },
      []
    );

    if (starts.length > 0) {
      let bef = '';
      let aft = '';
      if (
        endsToRemove.length === 1 &&
        (prevStarts.toString() !== starts.toString() ||
          // https://github.com/inokawa/remark-slate-transformer/issues/90
          (prevEnds.includes('italic') && ends.includes('bold'))) &&
        starts.length - endsToRemove.length === 0
      ) {
        while (textTemp.startsWith(' ')) {
          bef += ' ';
          textTemp = textTemp.slice(1);
        }
        while (textTemp.endsWith(' ')) {
          aft += ' ';
          textTemp = textTemp.slice(0, -1);
        }
      }
      let res: MdMark = {
        type: 'text',
        value: textTemp,
      };
      textTemp = '';
      starts
        .slice()
        .reverse()
        .forEach((k) => {
          const nodeParser = getSerializerByKey(k, options);

          if (nodeParser) {
            const node = nodeParser(cur, options) as any;
            res = {
              ...node,
              children: [res],
            };
          }

          switch (k) {
            case 'bold': {
              res = {
                children: [res],
                type: 'strong',
              };
              break;
            }
            case 'code': {
              let currentRes = res;
              while (
                currentRes.type !== 'text' &&
                currentRes.type !== 'inlineCode'
              ) {
                currentRes = currentRes.children[0] as MdMark;
              }
              currentRes.type = 'inlineCode';

              break;
            }
            case 'italic': {
              res = {
                children: [res],
                type: 'emphasis',
              };
              break;
            }
            case 'strikethrough': {
              res = {
                children: [res],
                type: 'delete',
              };
              break;
            }
          }
        });
      const arr: MdMark[] = [];
      if (bef.length > 0) {
        arr.push({ type: 'text', value: bef });
      }
      arr.push(res);
      if (aft.length > 0) {
        arr.push({ type: 'text', value: aft });
      }
      mdastTexts.push(...arr);
    }

    if (endsToRemove.length > 0) {
      endsToRemove.reverse().forEach((e) => {
        starts.splice(e.index, 1);
      });
    } else {
      mdastTexts.push({ type: 'text', value: textTemp });
      textTemp = '';
    }
  }
  if (textTemp) {
    mdastTexts.push({ type: 'text', value: textTemp });
    textTemp = '';
  }

  const mergedTexts = mergeTexts(mdastTexts);

  const flattenedEmptyNodes = mergedTexts.map((node) => {
    if (!hasContent(node)) {
      return { type: 'text', value: '' } as MdMark;
    }
    return node;
  });

  return flattenedEmptyNodes;
};

const hasContent = (node: MdMark): boolean => {
  if (node.type === 'inlineCode') {
    // inline has no children - no deeper search needed
    return node.value !== '';
  }

  if (node.type === 'text') {
    // inline has no children - no deeper search needed
    return node.value !== '';
  }

  if (node.children?.length > 0) {
    for (const child of node.children) {
      // all types other then emphasis are represented with some characters that can also be formatted
      if (
        child.type !== 'emphasis' &&
        child.type !== 'strong' &&
        child.type !== 'inlineCode' &&
        child.type !== 'delete' &&
        child.type !== 'text'
      ) {
        return true;
      }
      if (hasContent(child)) {
        return true;
      }
    }
  }
  return false;
};

const mergeTexts = (nodes: MdMark[]): MdMark[] => {
  const res: MdMark[] = [];
  for (const cur of nodes) {
    const last = res.at(-1);
    if (last && last.type === cur.type) {
      if (last.type === 'text') {
        last.value += (cur as typeof last).value;
      } else if (last.type === 'inlineCode') {
        last.value += (cur as typeof last).value;
      } else {
        last.children = mergeTexts(
          last.children.concat((cur as typeof last).children) as MdMark[]
        );
      }
    } else {
      if (cur.type === 'text' && cur.value === '') continue;
      res.push(cur);
    }
  }
  return res;
};
