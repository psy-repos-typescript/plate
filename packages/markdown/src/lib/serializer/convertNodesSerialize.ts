import {
  type Descendant,
  type TElement,
  type TText,
  getPluginKey,
  getPluginType,
  KEYS,
  TextApi,
} from 'platejs';

import type { unistLib } from '../types';
import type { SerializeMdOptions } from './serializeMd';

import { convertTextsSerialize } from './convertTextsSerialize';
import { listToMdastTree } from './listToMdastTree';
import { unreachable } from './utils';
import { getSerializerByKey } from './utils/getSerializerByKey';

export const convertNodesSerialize = (
  nodes: Descendant[],
  options: SerializeMdOptions
): unistLib.Node[] => {
  const mdastNodes: unistLib.Node[] = [];
  let textQueue: TText[] = [];

  const listBlock: TElement[] = [];

  for (let i = 0; i <= nodes.length; i++) {
    const n = nodes[i] as any;

    if (n && TextApi.isText(n)) {
      // Only add text nodes that pass the filtering
      if (shouldIncludeText(n, options)) {
        textQueue.push(n);
      }
    } else {
      if (textQueue.length > 0) {
        mdastNodes.push(
          ...(convertTextsSerialize(
            textQueue,
            options
          ) as any as unistLib.Node[])
        );
      }
      textQueue = [];
      if (!n) continue;

      // Skip this node if it doesn't pass the filtering
      if (!shouldIncludeNode(n, options)) {
        continue;
      }

      const pType = getPluginType(options.editor!, KEYS.p) ?? KEYS.p;

      if (n?.type === pType && 'listStyleType' in n) {
        listBlock.push(n);

        const next = nodes[i + 1] as TElement;
        const isNextIndent =
          next && next.type === pType && 'listStyleType' in next;

        if (!isNextIndent) {
          mdastNodes.push(listToMdastTree(listBlock as any, options));
          listBlock.length = 0;
        }
      } else {
        const node = buildMdastNode(n, options);

        if (node) {
          mdastNodes.push(node as unistLib.Node);
        }
      }
    }
  }

  return mdastNodes;
};

export const buildMdastNode = (node: any, options: SerializeMdOptions) => {
  const editor = options.editor!;

  let key = getPluginKey(editor, node.type) ?? node.type;

  if (KEYS.heading.includes(key)) {
    key = 'heading';
  }

  if (key === KEYS.olClassic || key === KEYS.ulClassic) {
    key = 'list';
  }

  const nodeParser = getSerializerByKey(key, options);

  if (nodeParser) {
    return nodeParser(node, options);
  }

  unreachable(node);
};

const shouldIncludeText = (
  text: TText,
  options: SerializeMdOptions
): boolean => {
  const { allowedNodes, allowNode, disallowedNodes } = options;

  // First check allowedNodes/disallowedNodes
  if (
    allowedNodes &&
    disallowedNodes &&
    allowedNodes.length > 0 &&
    disallowedNodes.length > 0
  ) {
    throw new Error('Cannot combine allowedNodes with disallowedNodes');
  }

  // Check text properties against allowedNodes/disallowedNodes
  for (const [key, value] of Object.entries(text)) {
    if (key === 'text') continue;

    if (allowedNodes) {
      // If allowedNodes is specified, only include if the mark is in allowedNodes
      if (!allowedNodes.includes(key) && value) {
        return false;
      }
    } else if (disallowedNodes?.includes(key) && value) {
      // If using disallowedNodes, exclude if the mark is in disallowedNodes
      return false;
    }
  }

  // Finally, check allowNode if provided
  if (allowNode?.serialize) {
    return allowNode.serialize(text);
  }

  return true;
};

const shouldIncludeNode = (
  node: TElement,
  options: SerializeMdOptions
): boolean => {
  const { allowedNodes, allowNode, disallowedNodes } = options;

  if (!node.type) return true;

  // First check allowedNodes/disallowedNodes
  if (
    allowedNodes &&
    disallowedNodes &&
    allowedNodes.length > 0 &&
    disallowedNodes.length > 0
  ) {
    throw new Error('Cannot combine allowedNodes with disallowedNodes');
  }

  if (allowedNodes) {
    // If allowedNodes is specified, only include if the type is in allowedNodes
    if (!allowedNodes.includes(node.type)) {
      return false;
    }
  } else if (disallowedNodes?.includes(node.type)) {
    // If using disallowedNodes, exclude if the type is in disallowedNodes
    return false;
  }

  // Finally, check allowNode if provided
  if (allowNode?.serialize) {
    return allowNode.serialize(node);
  }

  return true;
};
