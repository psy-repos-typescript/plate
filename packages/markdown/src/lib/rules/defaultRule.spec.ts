import { H1Plugin } from '@platejs/basic-nodes/react';
import { BoldPlugin } from '@platejs/basic-nodes/react';
import { createPlateEditor, ParagraphPlugin } from 'platejs/react';

import { deserializeMd } from '../deserializer';
import { MarkdownPlugin } from '../MarkdownPlugin';
import { serializeMd } from '../serializer';

describe('defaultRules', () => {
  it('should serialize custom keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'custom-h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
      ],
    });

    const result = serializeMd(editor, { value: nodes });
    expect(result).toMatchSnapshot();
  });

  it('should serialize custom mark', () => {
    const nodes = [
      {
        children: [
          { text: 'Paragraph' },
          { 'custom-bold': true, text: 'text' },
        ],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
        BoldPlugin.configure({
          node: { type: 'custom-bold' },
        }),
      ],
    });

    const result = serializeMd(editor, { value: nodes });
    expect(result).toMatchSnapshot();
  });

  it('should deserialize custom keys', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'custom-h1',
      },
      {
        children: [{ text: 'Paragraph' }],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
      ],
    });

    const result = deserializeMd(editor, '# Heading 1\nParagraph');
    expect(result).toEqual(nodes);
  });

  it('should deserialize custom mark', () => {
    const nodes = [
      {
        children: [{ text: 'Heading 1' }],
        type: 'custom-h1',
      },
      {
        children: [
          { text: 'Paragraph' },
          { 'custom-bold': true, text: 'text' },
        ],
        type: 'custom-p',
      },
    ];

    const editor = createPlateEditor({
      plugins: [
        MarkdownPlugin,
        H1Plugin.configure({
          node: { type: 'custom-h1' },
        }),
        ParagraphPlugin.configure({
          node: { type: 'custom-p' },
        }),
        BoldPlugin.configure({
          node: { type: 'custom-bold' },
        }),
      ],
    });

    const result = deserializeMd(editor, '# Heading 1\nParagraph**text**');
    expect(result).toEqual(nodes);
  });
});
