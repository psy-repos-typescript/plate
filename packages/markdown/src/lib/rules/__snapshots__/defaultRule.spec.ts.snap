// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultRules should serialize custom keys 1`] = `
"# Heading 1

Paragraph
"
`;

exports[`defaultRules should serialize custom mark 1`] = `
"Paragraph**text**
"
`;
